mlflow/models/evaluation/default_evaluator.py

```diff
@@ -1110,49 +1101,69 @@ def _log_custom_metric_artifact(self, artifact_name, raw_artifact, custom_metric
         artifact._load(artifact_file_local_path)
         return artifact
 
+    def _get_args_for_metrics(self, custom_metric, eval_df):
+        # deepcopying eval_df and builtin_metrics for each custom metric function call,
+        # in case the user modifies them inside their function(s).
+        eval_df_copy = eval_df.copy()
+        input_df = self.X.copy_to_avoid_mutation()
+        parameters = dict(inspect.signature(custom_metric.eval_fn).parameters)
+        eval_fn_args = {}
+        if "eval_df" in parameters:
+            eval_fn_args["eval_df"] = eval_df_copy
+            del parameters["eval_df"]
+            if "metrics" in parameters:
+                eval_fn_args["metrics"] = copy.deepcopy(self.metrics_values)
+            else:
+                metric_arg = list(parameters.values())[0].name
+                eval_fn_args[metric_arg] = copy.deepcopy(self.metrics)
+        else:
+            if "predictions" in parameters:
+                eval_fn_args["predictions"] = eval_df_copy["prediction"]
+                del parameters["predictions"]
+            if "targets" in parameters:
+                eval_fn_args["targets"] = eval_df_copy["target"]
+                del parameters["targets"]
+            if "metrics" in parameters:
+                eval_fn_args["metrics"] = copy.deepcopy(self.metrics_values)
+                del parameters["metrics"]
+            if "input" in parameters:
+                input_column_name = self.evaluator_config.get("input", "input")
+
+                if input_column_name in input_df.columns:
+                    eval_fn_args["input"] = input_df[input_column_name]
+                    del parameters["input"]
```

Comment:
I don't think we need to check the specific parameter names here. Following from https://github.com/mlflow/mlflow/pull/9746/files#r1340446020, if the function has 3 or more parameters, we know it has the new format. We also know that the new format parameters are strictly ordered: (predictions, targets, metrics). Accordingly, we can always pass these as the first 3 parameters. Any additional stuff that users want to add can go at the end (we can document this requirement that additional params come last).

Hopefully this simplifies things.

Reply:
I guess targets is not required. Its optional depending if we have passed the target in mlflow.evaluate. And so is metrics. We don't need _builtin_metrics values to compute these metrics. So its optional as well. So other than prediction, everything is optional I guess.